modules/numa-allocator/src/main/cpp/src/numa/numa_alloc.cpp

```diff
@@ -0,0 +1,196 @@
+            mask_ = numa_bitmask_alloc(size_);
+            copy_bitmask_to_bitmask(other.mask_, mask_);
+        }
+
+        void SetBit(size_t idx, bool val) {
+            if (idx < size_) {
+                if (val)
+                    numa_bitmask_setbit(mask_, idx);
+                else
+                    numa_bitmask_clearbit(mask_, idx);
+            }
+        }
+
+        void SetAll(bool val) {
+            if (val)
+                numa_bitmask_setall(mask_);
+            else
+                numa_bitmask_clearall(mask_);
+        }
+
+        bool GetBit(size_t idx) const {
+            if (idx < size_)
+                return numa_bitmask_isbitset(mask_, idx);
+            else
+                return false;
+        }
+
+        bool Equals(const BitSetImpl &other) const {
+            return numa_bitmask_equal(mask_, other.mask_);
+        }
+
+        size_t Size() const {
+            return size_;
+        }
+
+        bitmask *Raw() {
+            return mask_;
+        }
+
+        ~BitSetImpl() {
+            numa_bitmask_free(mask_);
+        }
+
+    private:
+        bitmask *mask_;
+        size_t size_;
+    };
+
+    BitSet::BitSet() {
+        p_impl_ = new BitSetImpl();
+    }
+
+    BitSet::BitSet(const BitSet &other) {
+        p_impl_ = new BitSetImpl(*other.p_impl_);
+    }
+
+    BitSet &BitSet::operator=(const BitSet &other) {
+        if (this != &other) {
+            BitSet tmp(other);
+            std::swap(this->p_impl_, tmp.p_impl_);
+        }
+        return *this;
+    }
+
+    bool BitSet::Get(size_t pos) const {
+        return p_impl_->GetBit(pos);
+    }
+
+    void BitSet::Set(size_t pos, bool value) {
+        p_impl_->SetBit(pos, value);
+    }
+
+    void BitSet::Set() {
+        p_impl_->SetAll(true);
+    }
+
+    void BitSet::Reset(size_t pos) {
+        p_impl_->SetBit(pos, false);
+    }
+
+    void BitSet::Reset() {
+        p_impl_->SetAll(false);
+    }
+
+    size_t BitSet::Size() const {
+        return p_impl_->Size();
+    }
+
+    bool BitSet::operator==(const BitSet &other) {
+        return this->p_impl_->Equals(*other.p_impl_);
+    }
+
+    bool BitSet::operator!=(const BitSet &other) {
+        return !(*this == other);
+    }
+
+    BitSet::~BitSet() {
+        delete p_impl_;
+    }
+
+    std::ostream &operator<<(std::ostream &os, const BitSet &set) {
+        os << '{';
+        for (size_t i = 0; i < set.Size(); ++i) {
+            os << set[i];
+            if (i < set.Size() - 1) {
+                os << ", ";
+            }
+        }
+        os << '}';
+        return os;
+    }
+
+    union region_size {
+        size_t size;
+        max_align_t a;
+    };
+
+    int NumaNodesCount() {
+        return numa_max_node() + 1;
+    }
+
+    template<typename Func, typename ...Args>
+    void *NumaAllocHelper(Func f, size_t size, Args ...args) {
+        auto ptr = static_cast<region_size *>(f(size + sizeof(region_size), args...));
```

Comment:
I think there should be a comment about the layout of a memory that is allocated.

Reply:
Yep,it should.

It is quite common trick to save size of allocated memory in first bytes of allocated region. But memory should be properly aligned, so I use union here.

Reply:
Fixed